# Semi Supervised (Sklearn) Demo

Developed by Jhonnatan Torres on Jan 12th, 2022

Version 1
___

**Goal**: Compare the performance of different *semi supervised* estimators available in [**scikit-learn**](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.semi_supervised)

## Data Preparation

Importing required libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.semi_supervised import LabelPropagation, LabelSpreading, SelfTrainingClassifier
from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import classification_report, accuracy_score

Using the *breast cancer* dataset for this demo

In [2]:
X, y = load_breast_cancer(as_frame=True, return_X_y=True)

In [3]:
X.describe()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
count   569.000000    569.000000  ...      569.000000               569.000000
mean     14.127292     19.289649  ...        0.290076                 0.083946
std       3.524049      4.301036  ...        0.061867                 0.018061
min       6.981000      9.710000  ...        0.156500                 0.055040
25%      11.700000     16.170000  ...        0.250400                 0.071460
50%      13.370000     18.840000  ...        0.282200                 0.080040
75%      15.780000     21.800000  ...        0.317900                 0.092080
max      28.110000     39.280000  ...        0.663800                 0.207500

[8 rows x 30 columns]

In [4]:
y.value_counts()

1    357
0    212
Name: target, dtype: int64

In [5]:
y.value_counts(normalize=True)

1    0.627417
0    0.372583
Name: target, dtype: float64

In [6]:
ss_index = X.sample(frac=0.5, random_state=1234).index

In [7]:
y_ss = y.copy()

In order to use the *semi supervised* estimators, the unknown labels should have the value **-1** and for this demo, it is assigned to **50%** of the overall records

In [8]:
y_ss.loc[ss_index] = -1

In [9]:
## The goal is to predict the labels with a "-1" value
y_ss.values

array([-1,  0,  0,  0,  0,  0, -1, -1,  0,  0,  0, -1,  0, -1,  0,  0,  0,
       -1,  0,  1, -1, -1, -1,  0, -1, -1,  0, -1,  0, -1,  0,  0,  0, -1,
       -1,  0, -1, -1,  0,  0, -1,  0,  0, -1, -1,  0,  1, -1,  1,  1, -1,
       -1,  1,  0, -1, -1,  0, -1, -1, -1,  1,  1,  0,  1, -1,  0, -1, -1,
        1, -1, -1, -1,  0, -1, -1,  0,  1, -1, -1,  1,  1,  1,  0,  0,  1,
       -1,  0,  0,  1, -1, -1, -1, -1, -1,  0, -1,  1, -1,  1, -1, -1, -1,
       -1,  1, -1,  0,  1,  1, -1, -1, -1, -1, -1, -1,  1, -1,  1,  0, -1,
        0,  1,  0, -1, -1, -1, -1,  0,  0, -1, -1,  1, -1,  0, -1, -1,  0,
        1,  1, -1,  1, -1, -1,  1, -1,  1,  1, -1,  1, -1, -1,  1,  1, -1,
        1,  1,  1, -1,  1,  1, -1, -1,  0, -1, -1,  0,  1, -1, -1, -1, -1,
        1, -1,  0, -1,  1,  1, -1,  0,  1,  1,  0, -1,  0,  1,  0, -1,  0,
       -1,  1, -1,  0,  1, -1,  0, -1,  1,  0,  0, -1,  0, -1,  0, -1,  0,
        1, -1, -1, -1,  1,  1,  0,  1,  0, -1, -1,  0,  1,  1, -1, -1, -1,
       -1, -1, -1,  1, -1

In [10]:
y_ss.value_counts()

-1    284
 1    176
 0    109
Name: target, dtype: int64

Using the index of the unknown labels to split the data into train and test datasets, the goal is compare the predictions against the actual values

In [11]:
X_train = X.drop(index=ss_index)
X_test = X.loc[ss_index]

y_train = y.drop(index=ss_index)
y_test = y.loc[ss_index]

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(285, 30) (284, 30) (285,) (284,)


In [12]:
X_train.head(10)

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
1         20.57         17.77  ...          0.2750                  0.08902
2         19.69         21.25  ...          0.3613                  0.08758
3         11.42         20.38  ...          0.6638                  0.17300
4         20.29         14.34  ...          0.2364                  0.07678
5         12.45         15.70  ...          0.3985                  0.12440
8         13.00         21.82  ...          0.4378                  0.10720
9         12.46         24.04  ...          0.4366                  0.20750
10        16.02         23.24  ...          0.2948                  0.08452
12        19.17         24.80  ...          0.3176                  0.10230
14        13.73         22.61  ...          0.3596                  0.14310

[10 rows x 30 columns]

In [13]:
X_test.head()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
531       11.670         20.02  ...          0.3206                  0.08950
166       10.800          9.71  ...          0.2090                  0.07699
485       12.450         16.41  ...          0.3231                  0.10340
66         9.465         21.01  ...          0.2878                  0.09211
220       13.650         13.16  ...          0.2380                  0.08718

[5 rows x 30 columns]

In [14]:
y_train.value_counts()

1    176
0    109
Name: target, dtype: int64

In [15]:
y_test.value_counts()

1    181
0    103
Name: target, dtype: int64

## Logistic Regression as a baseline

In [16]:
lr_pl = make_pipeline(StandardScaler(), 
                      LogisticRegression(C=0.1, class_weight='balanced', random_state=1234))

In [17]:
lr_pl.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('logisticregression',
                 LogisticRegression(C=0.1, class_weight='balanced',
                                    random_state=1234))])

In [18]:
lr_pl_preds = lr_pl.predict(X_test)

## Label Propagation

In [19]:
lp_pl = make_pipeline(StandardScaler(), 
                      LabelPropagation(kernel='knn', n_neighbors=13))

Please notice how the entire X was used to fit the estimator, in order to calculate the accuracy, the index of the "-1" labels will be used to compare the actuals against the predictions

In [20]:
lp_pl.fit(X, y_ss)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('labelpropagation',
                 LabelPropagation(kernel='knn', n_neighbors=13))])

In [21]:
lp_pl_preds = lp_pl.predict(X)

## Label Spreading

In [22]:
ls_pl = make_pipeline(StandardScaler(), 
                      LabelSpreading(kernel='knn', n_neighbors=13))

In [23]:
ls_pl.fit(X, y_ss)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('labelspreading',
                 LabelSpreading(kernel='knn', n_neighbors=13))])

In [24]:
ls_pl_preds = ls_pl.predict(X)

## Self Training Classifier

In [25]:
BASE_ESTIMATOR = LogisticRegression(C=0.1, class_weight='balanced', random_state=1234)
stc_pl = make_pipeline(StandardScaler(), 
                      SelfTrainingClassifier(base_estimator=BASE_ESTIMATOR))

In [26]:
stc_pl.fit(X, y_ss)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('selftrainingclassifier',
                 SelfTrainingClassifier(base_estimator=LogisticRegression(C=0.1,
                                                                          class_weight='balanced',
                                                                          random_state=1234)))])

In [27]:
stc_pl_preds = stc_pl.predict(X)

## Final Comparison
Comapring the performance of all the 4 estimators used in this demo

In [28]:
# Baseline Logistic Regression
print(classification_report(y_test, lr_pl_preds))

              precision    recall  f1-score   support

           0       1.00      0.90      0.95       103
           1       0.95      1.00      0.97       181

    accuracy                           0.96       284
   macro avg       0.97      0.95      0.96       284
weighted avg       0.97      0.96      0.96       284



In [29]:
# Baseline Label Propagation
print(classification_report(y_test, lp_pl_preds[ss_index]))

              precision    recall  f1-score   support

           0       1.00      0.83      0.90       103
           1       0.91      1.00      0.95       181

    accuracy                           0.94       284
   macro avg       0.95      0.91      0.93       284
weighted avg       0.94      0.94      0.94       284



In [30]:
# Baseline Label Spreading
print(classification_report(y_test, ls_pl_preds[ss_index]))

              precision    recall  f1-score   support

           0       1.00      0.83      0.91       103
           1       0.91      1.00      0.96       181

    accuracy                           0.94       284
   macro avg       0.96      0.92      0.93       284
weighted avg       0.95      0.94      0.94       284



In [31]:
# Baseline Self Training Classifier
print(classification_report(y_test, stc_pl_preds[ss_index]))

              precision    recall  f1-score   support

           0       0.99      0.91      0.95       103
           1       0.95      0.99      0.97       181

    accuracy                           0.96       284
   macro avg       0.97      0.95      0.96       284
weighted avg       0.97      0.96      0.96       284



In [32]:
print("Accuracy Baseline Logistic Regression: {:.3f}".format(accuracy_score(y_test, lr_pl_preds)))
print("Accuracy Label Propagation: {:.3f}".format(accuracy_score(y_test, lp_pl_preds[ss_index])))
print("Accuracy Label Spreading: {:.3f}".format(accuracy_score(y_test, ls_pl_preds[ss_index])))
print("Accuracy Self Training Classifier: {:.3f}".format(accuracy_score(y_test, stc_pl_preds[ss_index])))

Accuracy Baseline Logistic Regression: 0.965
Accuracy Label Propagation: 0.937
Accuracy Label Spreading: 0.940
Accuracy Self Training Classifier: 0.965


In [33]:
print("Mean CV Score Label Propagation: {:.3f}".format(cross_val_score(lp_pl, X, y_ss).mean()))
print("Mean CV Score Label Spreading: {:.3f}".format(cross_val_score(ls_pl, X, y_ss).mean()))
print("Mean CV Score Self Training Classifier: {:.3f}".format(cross_val_score(stc_pl, X, y_ss).mean()))

Mean CV Score Label Propagation: 0.478
Mean CV Score Label Spreading: 0.483
Mean CV Score Self Training Classifier: 0.489


*Low mean CV scores could be related to the "-1" labels used in these estimators*, predictions (1 or 0) are compared against "-1"

## Closure Comments

*   Mostly of the default parameters were used to fit the estimators
*   In terms of performance, both the Baseline (Logistic Regression) and the Self Training Classifier had the same accuracy score
*   Both Label Propagation and Label Spreading estimators had a pretty similar performance
*   At least for this particular dataset, and using 50% of the records as *unknown labels (-1)* the Label Propagation and Spreading estimators had a performance similar to the baseline, in addition, these estimators handled some imbalance (63% - 37% ratio) present in the dataset, without a *class weight* or *sample weight* parameter




